ecc/enclave_chaincode.go

```diff
@@ -174,6 +167,134 @@ func (t *EnclaveChaincode) setup(stub shim.ChaincodeStubInterface) pb.Response {
 	return shim.Success([]byte(enclavePkBase64))
 }
 
+// ============================================================
+// createenclave -
+// ============================================================
+func (t *EnclaveChaincode) createenclave(stub shim.ChaincodeStubInterface) pb.Response {
+	// TODO admin authentication?
+	// TODO check one time create?
+	logger.Debugf("createenclave")
+
+	// check if there is already an enclave
+	if t.enclave == nil {
+		return shim.Error("ecc: Enclave has already been initialized! Destroy first!!")
+	}
+
+	args := stub.GetStringArgs()
+
+	if len(args) != 3 {
+		errMsg := fmt.Sprintf("createenclave: unexpected params number")
+		logger.Errorf(errMsg)
+		return shim.Error(errMsg)
+	}
+
+	// grab attestation params
+	b64AttestationParams := args[1]
+	logger.Debugf("b64AttestationParams: %s", b64AttestationParams)
+	attestationParamsBytes, err := base64.StdEncoding.DecodeString(b64AttestationParams)
+	if err != nil {
+		errMsg := fmt.Sprintf("createenclave: b64AttestationParams: %s", err)
+		logger.Errorf(errMsg)
+		return shim.Error(errMsg)
+	}
+
+	// create HostParameters <<<
+	// grab host params
+	b64HostParams := args[2]
+	logger.Debugf("b64HostParams: %s", b64HostParams)
+	hostParamsBytes, err := base64.StdEncoding.DecodeString(b64HostParams)
+	if err != nil {
+		errMsg := fmt.Sprintf("createenclave: b64HostParams: %s", err)
+		logger.Errorf(errMsg)
+		return shim.Error(errMsg)
+	}
+
+	peerMSPID, err := shim.GetMSPID()
+	if err != nil {
+		errMsg := fmt.Sprintf("createenclave: peer mspid: %s", err)
+		logger.Errorf(errMsg)
+		return shim.Error(errMsg)
+	}
+	logger.Debugf("Peer MSPID: %s", peerMSPID)
```

Comment:
> I am afraid that relying on `shim.GetMSPID()` may cause running in difficulties. By default, starting fpc chaincode in "chaincode-as-a-service" mode, `CORE_PEER_LOCALMSPID` is not automatically set, hence, this call returns an error.
> Maybe we should just enforce the presents of `CORE_PEER_LOCALMSPID` when we start the chaincode server.

I thought from prior discussion (and also what is now documented in puml/protos as part of ercc verification), we agreed that the admin is what determines it and hence we would take it from the proposal/getCreator?  This should always exist?

Reply:
> I am afraid that relying on shim.GetMSPID() may cause running in difficulties. By default, starting fpc chaincode in "chaincode-as-a-service" mode, CORE_PEER_LOCALMSPID is not automatically set, hence, this call returns an error.
Maybe we should just enforce the presents of CORE_PEER_LOCALMSPID when we start the chaincode server.

This is addressed in/by our external builder https://github.com/hyperledger-labs/fabric-private-chaincode/blob/e91309789d9a934825013bee55e0d0de3ca49723/fabric/externalBuilder/bin/lib/common.sh#L101

> I thought from prior discussion (and also what is now documented in puml/protos as part of ercc verification), we agreed that the admin is what determines it and hence we would take it from the proposal/getCreator? This should always exist?

To be on the safe side, I have implemented both: get peer mspid, and get creator.

I am pretty sure that the MSPIDs in `GetMSPID` and (part of) `GetCreator` are different (if admin and peer are not under the same mspid). As in the proto we put `peer_msp_id`, it may not be accurate taking that with the `GetCreator`.

Reply:
@g2flyer can we add this to the externLauncher without the magic of `jq` easily? :)

Reply:
@mbrandenburger you mean `fabric/externalBuilder/chaincode_server` -- above quote seems from outdated master and probably refers to `fabric/externalBuilder/chaincode` .. -- note sure why would do anyting there via environment? Why not just put it into connection.json in the installed package (where all the other params are also are right now?)

Reply:
> I am pretty sure that the MSPIDs in `GetMSPID` and (part of) `GetCreator` are different (if admin and peer are not under the same mspid). As in the proto we put `peer_msp_id`, it may not be accurate taking that with the `GetCreator`.

In ERCC, verifying against creator seems the only easy way from a client-side processing perspective (not sure how you would get the signed peer endorsement in the cli to pass to ercc registration) and seems also appropriate as it brings an explicit (human) take-ownership; as there is "no privilege" and only "obligations" associate this take-ownership i don't see any security issue/reason why an (adversarial) admin of orgA would try to register an enclave at orgB . But checking both in ECC as you propose seems probably the best of both world

Reply:
> But checking both in ECC as you propose seems probably the best of both world

Oops, just realized that i misunderstood you: i thought you would cross-check in ECC that the two are the same but leave messages the same. Not sure whether adding creator to hostparams helps much (and currently ercc ignores it).   It really depends whether there is an easy way to get with chaincode_server an easyu way to make `shim.GetMSPID()' work. If that works then we can ignore creator in `ecc`. If not, then creator would be the source by which we could get it in that case, but in that case seems easier to just do it in ecc when filling out peer_msp_id rather than having all code which relies that HostParam field having to have code which has to look at both values, depending on which one exists?